File: NeetCode/Linked_List/Reverse_Nodes_In_K-Group/reverse__nodes__in__k-_group.py
from typing import Optional, List
from common.listnode import ListNode, build_list, print_list

from collections import deque


class Solution:
    def reverseKGroup(self, head: Optional[ListNode], k: int) -> Optional[ListNode]:
        dummy = ListNode(0)
        dummy.next = head
        ptr = dummy

        count = 0
        for _ in range(k):
            ptr = ptr.next
            if ptr == None:
                break
            count += 1
        if count != k:
            return head

        tmp1, tmp2 = ptr.next, dummy.next

        ptr.next = None
        dummy.next = self.reverseList(dummy.next)
        tmp2.next = self.reverseKGroup(tmp1, k)

        return dummy.next

    def reverseList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if head is None:
            return head

        ptr1 = head
        ptr2 = ptr1.next
        ptr1.next = None

        while ptr2:
            tmp = ptr2.next
            ptr2.next = ptr1
            ptr1 = ptr2
            ptr2 = tmp

        return ptr1


if __name__ == "__main__":
    sol = Solution()

    head = build_list([1, 2, 3, 4, 5, 6])
    k = 3

    print_list(sol.reverseKGroup(head, k))
    print("Running Solution...")

File: NeetCode/Linked_List/Reverse_Nodes_In_K-Group/Reverse_Nodes_In_K-Group.cpp
// Reverse_Nodes_In_K-Group.cpp
#include <algorithm>
#include <array>
#include <cassert>
#include <cctype>
#include <climits>
#include <cmath>
#include <deque>
#include <functional>
#include <iostream>
#include <map>
#include <numeric>
#include <queue>
#include <set>
#include <stack>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <utility>
#include <vector>

using namespace std;

struct ListNode
{
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution
{
public:
    ListNode *reverseKGroup(ListNode *head, int k)
    {
        ListNode dummy(0);
        dummy.next = head;
        ListNode *ptr = &dummy;

        // if an empty list, just return it
        if (dummy.next == nullptr)
        {
            return head;
        }
        // attempt to move ptr forward 3 times
        int cnt = 0;
        for (int i = 0; i < k; i++)
        {
            ptr = ptr->next;
            if (ptr == nullptr)
            {
                break;
            }
            cnt++;
        }
        // if couldn't get to the end, then the list is too short, return the head
        if (cnt != k)
        {
            return head;
        }

        // if the cnt reached 3, then we can process
        ListNode *tmp1 = ptr->next;
        ListNode *tmp2 = dummy.next;
        ptr->next = nullptr;
        dummy.next = reverseList(dummy.next);
        tmp2->next = reverseKGroup(tmp1, k);

        return dummy.next;
    }

    ListNode *reverseList(ListNode *head)
    {

        ListNode *curr = head;
        ListNode *prev = nullptr;
        ListNode *next = nullptr;

        while (curr != nullptr)
        {
            next = curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};
